File: components/esp_board_manager/peripherals/periph_dsi/periph_dsi.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# DSI peripheral config parser
VERSION = 'v1.0.0'

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'generators'))
from generators.utils.idf_version import get_idf_version

def get_includes() -> list:
    """Return list of required include headers for MIPI_DSI peripheral"""
    return [
        'esp_lcd_mipi_dsi.h',
    ]

def parse(name: str, config: dict) -> dict:
    """Parse MIPI_DSI peripheral configuration from YAML to C structure

    Args:
        name: Peripheral name
        config: Configuration dictionary from YAML

    Returns:
        Dictionary containing DSI configuration structure for esp_lcd_dsi_bus_config_t
    """
    try:
        # Get the config dictionary
        bus_cfg = config.get('config', {})

        # Get bus_id with validation
        bus_id = bus_cfg.get('bus_id', 0)
        if not isinstance(bus_id, int) or bus_id < 0:
            raise ValueError(f'Invalid bus_id: {bus_id}. Must be an integer >= 0.')

        # Get num_data_lanes
        num_data_lanes = bus_cfg.get('data_lanes', 2)
        if not isinstance(num_data_lanes, int) or num_data_lanes < 0:
            raise ValueError(f'Invalid data_lanes: {num_data_lanes}. Must be an integer >= 0.')

        # Get phy_clk_src with validation
        phy_clk_src = bus_cfg.get('phy_clk_src', 0)
        valid_phy_clk_srcs = [
            0,  # Set it to 0 and then let the driver configure it
            'MIPI_DSI_PHY_CLK_SRC_DEFAULT',
            'MIPI_DSI_PHY_CLK_SRC_PLL_F20M',
            'MIPI_DSI_PHY_CLK_SRC_PLL_F25M',
            'MIPI_DSI_PHY_CLK_SRC_RC_FAST'
        ]
        if phy_clk_src not in valid_phy_clk_srcs:
            raise ValueError(f'Invalid phy_clk_src: {phy_clk_src}. Valid values: {valid_phy_clk_srcs}')

        # Get lane_bit_rate_mbps
        lane_bit_rate_mbps = bus_cfg.get('lane_bit_rate_mbps', 0)
        if not isinstance(lane_bit_rate_mbps, int) or lane_bit_rate_mbps <= 0:
            raise ValueError(f'Invalid lane_bit_rate_mbps: {lane_bit_rate_mbps}. Must be an integer > 0.')

        flags = bus_cfg.get('flags', {})
        clock_lane_force_hs = bool(flags.get('clock_lane_force_hs', False))

        # Create base config
        struct_init = {
            'bus_id': bus_id,
            'num_data_lanes': num_data_lanes,
            'phy_clk_src': phy_clk_src,
            'lane_bit_rate_mbps': lane_bit_rate_mbps
        }
        if get_idf_version()[0] >= 6:
            struct_init['flags'] = {
                'clock_lane_force_hs': clock_lane_force_hs,
            }
        elif clock_lane_force_hs:
            print(
                f"YAML WARNING: DSI peripheral '{name}' field 'flags.clock_lane_force_hs' "
                'requires ESP-IDF v6.0 or later and will be ignored.'
            )

        result = {
            'struct_type': 'esp_lcd_dsi_bus_config_t',
            'struct_var': f'{name}_cfg',
            'struct_init': struct_init
        }
        return result

    except ValueError as e:
        # Re-raise ValueError with more context
        raise ValueError(f"YAML validation error in DSI peripheral '{name}': {e}")
    except Exception as e:
        # Catch any other exceptions and provide context
        raise ValueError(f"Error parsing DSI peripheral '{name}': {e}")

File: components/esp_board_manager/peripherals/periph_dsi/periph_dsi.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include "esp_err.h"
#include "esp_log.h"
#include "periph_dsi.h"
#include "esp_lcd_mipi_dsi.h"

static const char *TAG = "PERIPH_DSI";

int periph_dsi_init(void *cfg, int cfg_size, void **periph_handle)
{
    if (!cfg || !periph_handle || cfg_size != sizeof(esp_lcd_dsi_bus_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }
    esp_lcd_dsi_bus_handle_t handle = NULL;
    esp_err_t err = esp_lcd_new_dsi_bus((esp_lcd_dsi_bus_config_t *)cfg, &handle);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "MIPI DSI bus initialize failed: %s", esp_err_to_name(err));
        return -1;
    }
    ESP_LOGI(TAG, "MIPI DSI bus initialize success");
    *periph_handle = handle;
    return 0;
}

int periph_dsi_deinit(void *periph_handle)
{
    if (!periph_handle) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }
    ESP_LOGI(TAG, "MIPI DSI bus deinitialize");
    esp_lcd_dsi_bus_handle_t handle = (esp_lcd_dsi_bus_handle_t)periph_handle;
    esp_err_t err = esp_lcd_del_dsi_bus(handle);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "MIPI DSI bus del failed: %s", esp_err_to_name(err));
        return -1;
    }
    return 0;
}
